### PR TITLE
solution first draft

### PR DESCRIPTION
**Pull request by scottcreynolds**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,15 +9,115 @@ JavaScript Handlebars Templates Lab
...
+    <ul>
+      {{#each ingredients}}
+      <li>{{name}}</li>
+      {{/each}}
+    </ul>
+  </div>
+</script>
+```
+
+Keep in mind that the placeholder values within the double curly braces
```
Comment (pletcher) [inline README.md:38]: 👍

File: README.md [hunk 2]
```diff
@@ -9,15 +9,115 @@ JavaScript Handlebars Templates Lab
...
+</script>
+<script id="partial-template" type="text/x-handlebars-template">
+  <div>{{ name }}</div>
+</script>
+```
+```js
+Handlebars.registerPartial('namePartial', document.getElementById("partial-template").innerHTML)
+function renderMain() {
+  var template = Handlebars.compile(document.getElementById("main-template").innerHTML);
+  html = template({name: 'Gordon Ramsay'});
```
Comment (pletcher) [inline README.md]: Is this variable assumed to be declared in an outer scope?

Comment (scottcreynolds) [inline README.md]: no just an omission typo

File: README.md [hunk 3]
```diff
@@ -9,15 +9,115 @@ JavaScript Handlebars Templates Lab
...
+}
+```
+
+By default, the partial will receive the same context object as the
+template that calls it, so our partial has the same acces to `name` as
+the main template does.
+
+## Instructions
+
+1. Create a form template with an `id` of `recipe-form-template` that will be used to enter new recipes. Make the form submit with a `createRecipe()` function. Provide inputs for recipe `name`, `description`, and at least five `ingredients`. **Hint:** Get comfy collecting values with `getElementsByName()`.
```
Comment (pletcher) [inline README.md:99]: :heart: the hint

File: index.html [hunk 4]
```diff
@@ -6,6 +6,41 @@
   <body>
     <main id="main">
     </main>
+    <script src="handlebars.js"></script>
     <script src="index.js"></script>
+    <script id="recipe-template" type="text/x-handlebars-template">
+      <article>
+        <header><h2 id="nameHeader">{{name}}</h2></header>
+        <div>{{> recipeDetailsPartial }}</div>
```
Comment (pletcher) [inline index.html:14]: 🎸

File: index.html [hunk 5]
```diff
@@ -6,6 +6,41 @@
...
+        <footer><a href="#" onclick="displayEditForm();">Edit Recipe</a></footer>
+    </script>
+    <script id="recipe-details-partial" type="text/x-handlebars-template">
+      <p id="recipeDescription">{{description}}</p>
+      <ul>
+        {{#each ingredients}}
+        {{ displayIngredient this }}
```
Comment (pletcher) [inline index.html:21]: 🎹

File: index.js [hunk 6]
```diff
@@ -1 +1,68 @@
...
+  var template = Handlebars.compile(recipeTemplate)
+  document.getElementById("main").innerHTML = template(recipe)
+}
+
+function displayEditForm() {
+  var name = document.getElementById("nameHeader").innerText
+  var description = document.getElementById("recipeDescription").innerText
+  var ingredientsNodes = document.getElementsByName("ingredientsList")
+  var ingredients = []
+  for(var i=0;i<ingredientsNodes.length;i++) {
```
Comment (pletcher) [inline index.js:26]: What would you think of using `map()` here?

E.g, `var ingredients = Array.prototype.map.call(ingredientsNodes, node => node.innerText)`?

Hmmm... the Array bit is a little janky. `NodeList`s :(

File: index.js [hunk 7]
```diff
@@ -1 +1,68 @@
...
+  Handlebars.registerPartial('recipeFormPartial', document.getElementById("recipe-form-partial").innerHTML)
+}
+
+
+function init() {
+  //put any page initialization/handlebars initialization here
+  handlebarsSetup()
+  initForm()
+}
+document.addEventListener("DOMContentLoaded", function(event) {
```
Comment (pletcher) [inline index.js:66]: 💯

File: test/index-test.js [hunk 8]
```diff
@@ -6,13 +6,106 @@ const path = require('path')
...
+      expect(recipeTemplate.type).toBe("text/x-handlebars-template", "Template must be of type text/x-handlebars-template")
+      expect(recipeTemplate.innerHTML).toMatch(/{{\s?name\s?}}/)
+      expect(recipeTemplate.innerHTML).toMatch(/<a.*displayEditForm().*>Edit Recipe<\/a>/, "Template must have an 'Edit Recipe' link that calls 'displayEditForm()'")
+      expect(recipeTemplate.innerHTML).toMatch(/{{>\s?recipeDetailsPartial\s?}}/, "Template must render the recipeDetailsPartial")
+    })
+    it('has a recipe details partial template', () => {
+      var recipeDetailsPartial = document.getElementById("recipe-details-partial")
+      expect(recipeDetailsPartial).toExist("Must provide a template with an id of 'recipe-details-partial'")
+      expect(recipeDetailsPartial.type).toBe("text/x-handlebars-template", "Template must be of type text/x-handlebars-template")
+      expect(recipeDetailsPartial.innerHTML).toMatch(/{{\s?description\s?}}/)
```
Comment (pletcher) [inline test/index-test.js:45]: I kinda like this way of testing -- normally looking for strings in student code is a bit janky, but this is nice/makes sure they do the right thing™.

**Comment by pletcher** [comment]
LGTM! I just fixed that minor `var` thing. Ready to :ship: when you are!

File: test/index-test.js (hunk 8 above)
Comment (scottcreynolds) [inline test/index-test.js:45]: yeah it goes against my normal testing principles but seemed like a good compromise to make sure at least they're doing the basic template stuff. took a while to land on something i liked for testing this one.

**Pull request merged**